### PR TITLE
[ADT] Make set_subtract more efficient when subtrahend is larger (NFC)

### PR DESCRIPTION
**Pull request by teresajohnson**
If the subtrahend is larger, iterate the minuend set instead.

Noticed when subtracting a large set from a number of other smaller
sets for an upcoming MemProf change, this change makes that much faster.

I subsequently found a couple of callsites in one file that were calling
set_subtract with a vector subtrahend, which doesn't have the "count()"
interface. Add a separate helper for subtracting a vector.

### CODE REVIEW

File: llvm/include/llvm/ADT/SetOperations.h [hunk 1]
```diff
@@ -92,9 +92,27 @@ S1Ty set_difference(const S1Ty &S1, const S2Ty &S2) {
...
+/// Selects the set to iterate based on the relative sizes of A and B for better
+/// efficiency.
+///
 template <class S1Ty, class S2Ty> void set_subtract(S1Ty &S1, const S2Ty &S2) {
+  if (S1.size() < S2.size()) {
+    for (typename S1Ty::iterator SI = S1.begin(), SE = S1.end(); SI != SE; ++SI)
+      if (S2.count(*SI))
+        S1.erase(SI);
+    return;
+  }
```
Comment (kazutakahirata) [inline llvm/include/llvm/ADT/SetOperations.h:115]: I wonder if we could teach `set_subtract` to use the new path `S1.size() < S2.size()` when either `S2Ty::contains` or `S2Ty::find` is available:

```suggestion
  using ElemTy = decltype(*S1.begin());
  if constexpr (detail::HasMemberContains<S2Ty, ElemTy> ||
                detail::HasMemberFind<S2Ty, ElemTy>) {
    if (S1.size() < S2.size()) {
      for (typename S1Ty::iterator SI = S1.begin(), SE = S1.end(); SI != SE; ++SI)
        if (llvm::is_contained(S2, *SI))
          S1.erase(SI);
      return;
    }
  }
```

This way, we don't have to have `set_subtract_vec`.  If `S2Ty` is a vector, the `constexpr if` would automatically drop the new path.  I'm using `detail::HasMemberContains<S2Ty, ElemTy>` and `detail::HasMemberFind<S2Ty, ElemTy>` as a heuristic to see if `S2Ty` offers a sublinear lookup like O(1) or O(log(n)).  Calling `is_contained` without the `constexpr if` would be a disaster, invoking a linear look up when `S2Ty` is a vector.

@kuhar Any thoughts here?  I recall you generalized `is_contained` a while ago at 0eaacc25bb98f50cc98bab5f6ef8d6d67e112317.

Comment (kuhar) [reply llvm/include/llvm/ADT/SetOperations.h:115]: This makes sense to me in the absence of more standardized type traits we could use.

Comment (teresajohnson) [reply llvm/include/llvm/ADT/SetOperations.h]: Great tip! Done.

File: llvm/include/llvm/ADT/SetOperations.h [hunk 2]
```diff
@@ -94,7 +94,24 @@ S1Ty set_difference(const S1Ty &S1, const S2Ty &S2) {
...
+/// Selects the set to iterate based on the relative sizes of A and B for better
+/// efficiency.
+///
 template <class S1Ty, class S2Ty> void set_subtract(S1Ty &S1, const S2Ty &S2) {
+  using ElemTy = decltype(*S1.begin());
+  // A couple callers pass a vector for S2, which doesn't support count(), and
+  // wouldn't be efficient if it did. In the absence of a more direct check,
+  // ensure the type supports the contains or find interfaces.
+  if constexpr (detail::HasMemberContains<S2Ty, ElemTy> ||
+                detail::HasMemberFind<S2Ty, ElemTy>) {
```
Comment (nikic) [inline llvm/include/llvm/ADT/SetOperations.h:106]: I think this should only check for contains, and the count call below should be adjusted to use contains instead to match.

Comment (kazutakahirata) [reply llvm/include/llvm/ADT/SetOperations.h:106]: I think the `detail::HasMemberContains` and `detail::HasMemberFind` checks should go with `is_contained` in place of `count` below.  This way, `is_contained` will pick `C.contains(E)` or `C.find(E) != C.end()` for you depending on the availability.

Comment (teresajohnson) [reply llvm/include/llvm/ADT/SetOperations.h]: I went ahead and changed it to just check and use contains

**Review by kazutakahirata — APPROVED** [review]
LGTM.  Thanks!

**Review by kuhar — APPROVED** [review]

**Pull request merged**